changelogs/unreleased/uri-library-extension.md

```diff
@@ -0,0 +1,8 @@
+## feature/uri
+
+* Implemented new method `parse_many` in tarantool uri library,
```

Comment:
Why would anybody need that in their application? Why is it public?

Reply:
This was requested by @alyapunov:
- We need this functionality to parse uris passed via environment.
- Users might need it for the same reason (e.g. parse uris from environment, then add some parameters, then pass to net.box).
- Since we already implemented this function, we can expose it to everyone - it wouldn't hurt.
- Also, it makes it easier for us to test this function in case it's a part of the public library API.

Reply:
* Why @alyapunov needs that to be public?
* Is there a single user who actually asked for that? So that it would be needed in public. To parse this 'uri set' exactly the same way as you do the users can simply do this:
```Diff
diff --git a/src/lua/uri.lua b/src/lua/uri.lua
index 98d5aa7ce..5290e2cca 100644
--- a/src/lua/uri.lua
+++ b/src/lua/uri.lua
@@ -110,18 +110,11 @@ local function parse_many(str)
     if str == nil then
         error("Usage: uri.parse_many(string)")
     end
-    local uri_set_buf = uri_set_stash_take()
-    if builtin.uri_set_create(uri_set_buf, str) ~= 0 then
-        uri_set_stash_put(uri_set_buf)
-        return nil
-    end
-    local result = {}
-    for i = 0, uri_set_buf.uri_count - 1 do
-        result[i + 1] = parse_uribuf(uri_set_buf.uris[i])
+    local urls = str:split(',')
+    for i = 1, #urls do
+        urls[i] = parse(urls[i]:strip())
     end
-    builtin.uri_set_destroy(uri_set_buf)
-    uri_set_stash_put(uri_set_buf)
-    return result
+    return urls
 end
```
These are 5 lines of functional code. Including `end` and `return`. What exactly does `uri_set` do better than this code? The only difference I saw was that your code considers multiple `,,` as an error, while `string.split` will drop excess commas.
* *"Since we already implemented this function, we can expose it to everyone - it wouldn't hurt."* - oh no, not again. Leaving code just because it is already done is not a way to go. It would actually hurt, because 1) is adds more public code to support and to keep compatibility with, 2) it tells other devs they can do tons of code no-one asked for and it will be silently submitted. Just because it is done already.
* *"Also, it makes it easier for us to test this function in case it's a part of the public library API."* - how is it simpler? We don't need code to be public in order to test it. For example, Raft library is not public and yet it is unit tested just fine. In Lua as well we have tons of internal code which is not public and yet is covered by Lua tests.

With that said, I don't see any sense to expose `uri_set` to Lua at all. Not to public neither to private API. In Lua it is replaceable by a plain Lua array like I have shown in the diff above. Lets ask @Mons to resolve the conflict. Since it is product manager's responsibility to deal with public API changes anyway.

Reply:
uri.set works not only with strings, separated by commas but also with table of tables.
```lua
uri_set = uri.parse_many("uri?x=y", params = { z = "k" })
```
And so on, same as "listen" and "replication" cfg setttings.

Reply:
There is no such thing, as "many uris" in URI RFC. No libraries have such method. No strict rules for the following thing.
So, don't ever make it public.

Reply:
Also consult, please, https://datatracker.ietf.org/doc/html/rfc3986 for the usage of different symbols, like `,` (see sub-delims)

Reply:
I want tarantool to be user-friendly, to be not so high threshold for entry. In terms of config that means a user can correctly configure tarantool without (or less) checking the documentation.

I'm talking about developers that try to use tarantool. I understand than in production installations config is usually set by robots or almost by robots. But, for example, for me, it's important to write a line in script and get the result.

I'm not sure how exactly users would try to set, for example, cfg.listen. Perhaps there are several groups the would expect different approaches. I'm know for sure that it's quite expensive to make such a demand investigation (@Mons, it that your words?). Fortunately it's not so expensive to support different approached to comply different possible users' demands.

Since we need to understand a variety of URL syntax in different places, that parsing (or, I would say, canonizing) must be implemented in a sort of library. I mean generally split+strip is not enough. Perhaps we can make internal URI library that is used by tnt developers along with public URL library that is used by tnt users, but IMO that's an overkill.

>There is no such thing, as "many uris" in URI RFC.
There are many URIs in real life. box.cfg.listen must accept a list URIs. There's no conflict for URI library to have functions that handles URI lists.